### ISSUE TITLE
UFBX Imports Empties with 100x Scale & Wrong Rotation

### ISSUE THREAD

**Issue by Uradamus** [godotengine/godot#90314]
### Tested versions

v4.3.dev4.official [df78c0636]

### System information

Godot v4.3.dev4 - Manjaro Linux #1 SMP PREEMPT_DYNAMIC Thu Apr  4 20:32:38 UTC 2024 - X11 - GLES3 (Compatibility) - AMD Radeon RX 570 Series (radeonsi, polaris10, LLVM 16.0.6, DRM 3.54, 6.6.25-1-MANJARO) () - AMD Ryzen 5 5600G with Radeon Graphics (12 Threads)

### Issue description

Importing FBX files that contain empties with the new UFBX feature results in empties that are 100x bigger than they should be and with the wrong rotation.

I was testing to see if I could set up scenes in Blender that used empties as instance placeholders that retain their transformations upon import in Godot as node3d nodes that I can parent my instances to. It works fine with gltf, but fbx has the above mentioned issues.

I exported the instance_test scene as both file types with default export settings in Blender and created a new inherited scene for each, then instanced the instance_marker scene into each empty on each version of the scene for the MRP below. So the gltf.tscn shows the expected results while the ufbx.tscn shows the issues being experienced.

### Steps to reproduce

Drop an fbx containing empties into a project and inspect the transforms on the resulting node3d.

### Minimal reproduction project (MRP)

[ufbx_test.zip](https://github.com/godotengine/godot/files/14894397/ufbx_test.zip)

**Comment by fire** [comment]
@bqqbarbhg Any ideas?

**Comment by bqqbarbhg** [comment]
Blender stores by default all objects at 100x scale, using 1cm units, in Godot we use `UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY` unconditionally, which scales meshes, but leaves nodes be:

https://github.com/godotengine/godot/blob/e5b4ef8e9522e950033cbece39a31a4a76da19c1/modules/fbx/fbx_document.cpp#L1956

Using `UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS` would scale the nodes, but then cause other kinds of scaling issues in files exported by other exporters. This could be added as an import option, ie. allowing users to choose between `UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY` and `UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS` with better names, though I think last time the general consensus was that it's too much complexity.

Fortunately, Blender allows exporting with the setting "Apply Scalings" set to "FBX Units Scale", which will export the file as 1m units, removing any funky transforms that appear.

Anyways, I'll take a peek at what FBX2glTF does here and see if there's some clean solution, but I don't know if there is one.

**Comment by bqqbarbhg** [comment]
Looks like FBX2gLTF behaves exactly the same way here, so I'd say ufbx does the "correct" thing here.

Exposing the `UFBX_SPACE_CONVERSION_ADJUST_TRANSFORMS/UFBX_SPACE_CONVERSION_MODIFY_GEOMETRY` as some user-friendly option like `Unit Scaling: Node Transforms / Geometry` would be one option.

But as I said for Blender this can also be avoided on the export side by using "FBX Units Scale".

**Comment by Uradamus** [comment]
Just to confirm - I tested exporting the fbx with that "apply transform" option checked and it indeed fixed both the scaling and rotation issues for the generated nodes after reimporting and creating a new inherited scene.

So one option would be to treat this as a documentation issue and make sure this edge case and its solution shows up in the documentation. It wouldn't be wise to recommend using that option outside of this use case though, since applying transforms can screw up armatures and animations from what I recall. But still, I can't see really any use cases that would need both empties and armatures in the same FBX, but you never know.

**Comment by Uradamus** [comment]
[instance-test.zip](https://github.com/godotengine/godot/files/14908496/instance-test.zip)
Here is the latest export (using the apply transform option) of the fbx that had issues in the MRP on its own, in case any of you would like to further test with it.

**Comment by fire** [comment]
This is not very visible, but we placed the documentation here. 

https://docs.godotengine.org/en/latest/classes/class_fbxdocument.html#description

**Comment by Uradamus** [comment]
I think perhaps the wording in the Blender exporter might have changed, if that documentation was accurate at one point. Because right now there are no options labeled "FBX Unit Scale" that I can see. I assume it is equivalent to the "Apply Transform" option I mentioned above, since that is the option that fixed this issue.

I haven't used FBX export much though, so not totally sure. There is a chance it could also be something in the commercial Better FBX Importer & Exporter add-on that is pretty popular, and perhaps the author of that page didn't realize they were looking at the wrong exporter, but that's purely speculation. But I know that add-on does have a field that lets you adjust the base FBX units.

Included is a screenshot of the default exporter settings. I've underlined the option in question in red. It's unchecked by default and has the warning about potential issues with armatures/animations I mentioned earlier in the form of that caution symbol that gives some context when you hover over it.
![Screenshot_20240408_225511](https://github.com/godotengine/godot/assets/4281292/ae863fc3-244f-4868-9d15-ce2470e97ec4)

**Comment by bqqbarbhg** [comment]
The setting is in the "Apply Scalings" dropdown as I mentioned in my first reply:

> Fortunately, Blender allows exporting with the setting "Apply Scalings" set to "FBX Units Scale", which will export the file as 1m units, removing any funky transforms that appear.

Using this is also fully safe (as far as ufbx is concerned), as the only thing it does is to set the file unit scale to be meters, and skips scaling all the nodes by 100. I think the default is what it is for broken importers that ignore the FBX file units, as centimeters are the de-facto standard unit for FBX.

**Comment by bqqbarbhg** [comment]
It's here, a bit above the "Apply Transform" setting:

![image](https://github.com/godotengine/godot/assets/3875461/6cdd41e5-6a13-4eda-a0c4-6f2300390f93)

**Comment by Uradamus** [comment]
I have no idea how I wasn't seeing that, lol.

**Comment by bqqbarbhg** [comment]
Oh also of note, using that will not remove the 90 degree rotation from the nodes, as that's needed to convert the scene from Blender's Z-up to Godot's Y-up. "Apply Transform" will probably get rid of that at the cost of messing up armatures.

**Comment by Uradamus** [comment]
Ya, just noticed that. Though the weird part about that is that the Blender FBX exporter defaults to -z-forward and y-up. As can be seen in both of our screenshots.

**Comment by bqqbarbhg** [comment]
Yeah if you export with Y-up in Blender, Blender itself will add the 90 degree rotation to nodes. If you export Z-up the file metadata will contain the information that it is Z-up, but nodes will have no extra transforms. Unfortunately, when loading, ufbx will add the transforms back to make the file look correct :D

So in essence, changing the export space just changes where the conversion is done.

**Comment by Uradamus** [comment]
Ya, so after trying a ton of different combos, it seems the "apply transform" is the only one that fixes the rotation issue. It also technically can fix the scale issue as well, since as long as that is checked it doesn't seem to matter what the "Apply Scalings" is set to. Though the FBX Units Scale seems like a sensible default none the less for that option.

**Comment by fire** [comment]
As far as I know, this is an intentional preference compared to the Blender export of fbx, not sure if we can do anything about it. As far as I know there is no visual difference.

How can we best resolve this?

**Comment by fire** [comment]
As far as I know we're not planning on fixing this issue.

There are documentation workarounds.

**Comment by ChildLearningClub** [comment]
Creating Convex and Trimesh collisions from the imported .fbx files with the scale applied results in GodotPhysics issues, with  the collisions themselves scaled to match the scaled .fbx files utilizing Godot's built in "Create Collsion Shape". 

Using the MeshDataTool is required to create collision shapes from these scaled meshes with a scale of 1:1 to work with the physics engine. I think that it would be ideal that the FBX importer does this and reset the scale to 1:1 on the meshes? 

```
	var mesh_node_instances: Array[Node] = imported_fbx_node.find_children("*", "MeshInstance3D", true, false)
	for mesh_node: MeshInstance3D in mesh_node_instances:

		var mesh_instance_3d_node = MeshInstance3D.new()
		var expanded_mesh = ArrayMesh.new()
		
		var original_mesh = mesh_node.mesh as ArrayMesh
		
		# Create a new ArrayMesh and MeshDataTool
		var mdt = MeshDataTool.new()
		var surface_count = original_mesh.get_surface_count()
		
	# Iterate through each surface in the mesh
		for surface_index in range(surface_count):
			# Create MeshDataTool from the current surface
			mdt.create_from_surface(original_mesh, surface_index)

			var vertex_count = mdt.get_vertex_count()

			# Expand the vertices for this surface
			for i in range(vertex_count):
				var vertex = mdt.get_vertex(i)
				# Expand the vertex by the expansion factor
				vertex *= mesh_node.scale
				mdt.set_vertex(i, vertex)

			# If the expanded mesh is empty, commit the first surface, otherwise append the surface
			if surface_index == 0:
				expanded_mesh.clear_surfaces()
				mdt.commit_to_surface(expanded_mesh, surface_index)
			else:
				mdt.commit_to_surface(expanded_mesh, surface_index)
```

**Comment by fire** [comment]
Although this bug is in the FBX importer, I think this step should be applied in the Scene Importer.

**Comment by fire** [comment]
> ideal that the FBX importer does this and reset the scale to 1:1 on the meshes

I haven't looked at FBX baking in a long time, but typically if we do this feature we break armatures.

**Comment by AyyZerrAsa** [comment]
> Ah there's actually some inconsistencies with applying the node transform in the importer's existing code. `get_collision_shapes_transform` in the importer needs to be multiplied with the node transform. Will fix this in my PR - this does not touch the "Create Collision Shape" UI option that's mentioned above, however.

### TL;DR My fbx version of the mesh happened to be smaller than the min possible size of a collision box axis.

For example:
A 5cm mesh with a node scale of 1 would create a box collision shape of 0.05 units in size.
A 5cm mesh with a node scale of 100 would create a box collision shape of 0.0005 units in size, which hit the min limit of **0.001** and returns 0.001.

Will look at lowering this minimum to **FLT_EPSILON** rather than **0.001** (which seems a bit arbitrary) - this should cover small enough shapes when working with a 100x scale.

Otherwise, Blender's default scaling is annoyingly obfuscated and also not obvious in Godot's scene importer. 

### Perhaps offering transparency for transforms within the importer would clear up some confusion around scaling, especially as unless the scene is instantiated/code is written, there doesn't seem to be an easy way of obtaining transform data.

**Comment by fire** [comment]
The main mission of the fbx importer is to support Maya and other standard fbxsdk files. We also support blender fbx because it is a major file creator but we had to special case it.

Also there's not that much space to describe the components of the fbx transform.

```c++
// For a better reference you can check editor_scene_importer_fbx.h 
// references: GenFBXTransform / read the data in
// references: ComputePivotTransform / run the calculation 
// This is the local pivot transform for the node, not the global transforms
Transform ComputePivotTransform(
		Transform chain[TransformationComp_MAXIMUM],
		Transform &geometric_transform) {

	// Maya pivots
	Transform T = chain[TransformationComp_Translation];
	Transform Roff = chain[TransformationComp_RotationOffset];
	Transform Rp = chain[TransformationComp_RotationPivot];
	Transform Rpre = chain[TransformationComp_PreRotation];
	Transform R = chain[TransformationComp_Rotation];
	Transform Rpost = chain[TransformationComp_PostRotation];
	Transform Soff = chain[TransformationComp_ScalingOffset];
	Transform Sp = chain[TransformationComp_ScalingPivot];
	Transform S = chain[TransformationComp_Scaling];

	// 3DS Max Pivots
	Transform OT = chain[TransformationComp_GeometricTranslation];
	Transform OR = chain[TransformationComp_GeometricRotation];
	Transform OS = chain[TransformationComp_GeometricScaling];

	// Calculate 3DS max pivot transform - use geometric space (e.g doesn't effect children nodes only the current node)
	geometric_transform = OT * OR * OS;
	// Calculate standard maya pivots
	return T * Roff * Rp * Rpre * R * Rpost.inverse() * Rp.inverse() * Soff * Sp * S * Sp.inverse();
}
```

https://github.com/RevoluPowered/FBX-Document-Notes/blob/master/index.md
